File: ports/csharp/src/gui/kernel/BcsInputArbitrator.cs
using System;
using Bcs.Core.Kernel;
using Bcs.Gui.Widgets;

namespace Bcs.Gui.Kernel
{
    public enum DeviceType
    {
        Pointer = 0,
        Keyboard,
        Touch,
        Tablet,
        NumDeviceTypes
    }

    public class BcsInputArbitrator : BcsObject
    {
        private static readonly Lazy<BcsInputArbitrator> _instance = new Lazy<BcsInputArbitrator>(() => new BcsInputArbitrator());
        public static BcsInputArbitrator Instance => _instance.Value;

        private readonly int[] _deviceCounts = new int[(int)DeviceType.NumDeviceTypes];
        private readonly object _lock = new object();

        private BcsInputArbitrator() : base(null) {}

        public int DeviceCount(DeviceType type)
        {
            lock (_lock)
            {
                if (type >= 0 && type < DeviceType.NumDeviceTypes)
                {
                    return _deviceCounts[(int)type];
                }
                return 0;
            }
        }

        public void SetDeviceCount(DeviceType type, int count)
        {
            lock (_lock)
            {
                if (type >= 0 && type < DeviceType.NumDeviceTypes)
                {
                    _deviceCounts[(int)type] = count;
                }
            }
        }

        public bool RouteEvent(BcsEvent e, BcsWidget target)
        {
            if (target != null)
            {
                return target.Event(e);
            }
            return false;
        }
    }
}

File: ports/rust/src/gui/kernel/bcs_input_arbitrator.rs
use std::sync::{Arc, Mutex, OnceLock};
use crate::core::kernel::bcs_object::BcsObject;
use crate::core::kernel::bcs_event::BcsEventT;
use crate::gui::widgets::bcs_widget::BcsWidget;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeviceType {
    Pointer = 0,
    Keyboard,
    Touch,
    Tablet,
    NumDeviceTypes,
}

pub struct BcsInputArbitrator {
    pub base: Arc<BcsObject>,
    device_counts: Mutex<[i32; 4]>,
}

static INSTANCE: OnceLock<Arc<BcsInputArbitrator>> = OnceLock::new();

impl BcsInputArbitrator {
    pub fn instance() -> Arc<BcsInputArbitrator> {
        INSTANCE.get_or_init(|| {
            Arc::new(Self {
                base: BcsObject::new(),
                device_counts: Mutex::new([0; 4]),
            })
        }).clone()
    }

    pub fn device_count(&self, t: DeviceType) -> i32 {
        let counts = self.device_counts.lock().unwrap();
        counts[t as usize]
    }

    pub fn set_device_count(&self, t: DeviceType, count: i32) {
        let mut counts = self.device_counts.lock().unwrap();
        counts[t as usize] = count;
    }

    pub fn route_event(&self, _event: &dyn BcsEventT, _target: Option<Arc<BcsWidget>>) -> bool {
        // Mock routing
        false
    }
}
